File: Algorithms/src/2022. Convert 1D Array Into 2D Array/Code/solution.js
/**
 * This function converts a 1D array into a 2D array with specified dimensions.
 *
 * @param {number[]} original - The original 1D array that needs to be converted.
 * @param {number} m - The number of rows desired in the 2D array.
 * @param {number} n - The number of columns desired in the 2D array.
 * @return {number[][]} - The resulting 2D array, or an empty array if the conversion is not possible.
 */
var construct2DArray = function (original, m, n) {
  // Step 1: Check if the total number of elements in the original array
  // matches the required number of elements in the 2D array (m * n).
  // If not, it's impossible to form the desired 2D array, so return an empty array.
  if (original.length !== m * n) return [];

  // Step 2: Initialize an empty array to hold the rows of the 2D array.
  let result = [];

  // Step 3: Loop over the number of rows (m) to create each row of the 2D array.
  for (let i = 0; i < m; i++) {
    // For each row, slice a portion of the original array that corresponds to the row.
    // The slice starts at index i * n and ends just before index (i + 1) * n.
    // This gives us a sub-array with 'n' elements, representing a single row.
    result.push(original.slice(i * n, (i + 1) * n));
  }

  // Step 4: Return the fully constructed 2D array.
  return result;
};

File: Algorithms/src/2022. Convert 1D Array Into 2D Array/Code/solution.go
func construct2DArray(original []int, m int, n int) [][]int {
    // Step 1: Check if the total number of elements in the original array
    // matches the required size for the 2D array (m * n).
    // If not, return an empty 2D array because it's not possible to form
    // a valid 2D array with the given dimensions.
    if len(original) != m * n {
        return [][]int{} // Return an empty array if the sizes don't match
    }
    
    // Step 2: Initialize the 2D array 'result' with 'm' rows.
    // Each row will be a slice of integers (initially nil).
    result := make([][]int, m)
    
    // Step 3: Fill each row of the 2D array by slicing the original array.
    // Loop through each row index from 0 to m-1.
    for i := 0; i < m; i++ {
        // For the i-th row, take a slice from the original array starting at
        // index 'i*n' to index '(i+1)*n'. This slice forms the i-th row of the 2D array.
        result[i] = original[i*n : (i+1)*n]
    }
    
    // Step 4: Return the fully constructed 2D array.
    return result
}
